# Decision Tree

### Import Library

In [2]:
import pydotplus 
from IPython.display import Image
import numpy as np
import pandas
import matplotlib.pyplot as plt
from sklearn import tree
import os
from collections import defaultdict
import _pickle as cPickle
from scipy.stats import kurtosis, skew
%matplotlib notebook

### Data Preprocessing
- Two classes

In [3]:
def preprocessing(data):
    result = []
    mean = np.mean(data)
    median = np.median(data)
    maximum = np.amax(data)
    minimum = np.amin(data)
    std_dev = np.std(data)
    var = np.var(data)
    ran = np.ptp(data)
    skewness = skew(data)
    kurto = kurtosis(data)
    result=[mean, median, maximum, minimum,
                   std_dev, var, ran, skewness, kurto]
    return result

In [4]:
dirname = './data_preprocessed_python'
filenames = os.listdir(dirname)
data_dic = defaultdict(lambda:[])
for i, filename in enumerate(filenames):
    full_filename = os.path.join(dirname, filename)
    x = cPickle.load(open(full_filename, 'rb'), encoding='ISO-8859-1')
    data_dic[filename[:-4]] = x

In [5]:
new_data_dic = defaultdict(lambda:defaultdict(lambda:[]))
for key in data_dic.keys():
    par = int(key[1:])
    
    for i, vid in enumerate(data_dic[key]['data']):
        new_data_dic[key]['data'].append([])
        for chnn in vid[:-8]:
            l =preprocessing(np.array(chnn))
            for batch in range(0,10):
                l = l+preprocessing(np.array(chnn[batch*807:max((batch+1)*807, 8064)]))
            new_data_dic[key]['data'][i].append(np.array([par, i+1]+l))
    for i,  vid in enumerate(data_dic[key]['labels']):
        l2 = []
        new_data_dic[key]['labels'].append([])
        for value in vid:
            if value>=5:
                l2.append(1)
            else:
                l2.append(0)
        new_data_dic[key]['labels'][i] += l2

### Set Train_data and Test_data

In [6]:
train_data = []
train_label = []
test_data = []
test_label = []
for i in new_data_dic:
    print("process {}".format(i))
    if i != 's02':
        
        for j in range(len(new_data_dic[i]['labels'])):
            temp = np.array(new_data_dic[i]['data'][j]).reshape(-1, 32*101)[0]
            train_data.append(temp)
            train_label.append(new_data_dic[i]['labels'][j])
    else:
        for j in range(len(new_data_dic[i]['labels'])):
            temp = np.array(new_data_dic[i]['data'][j]).reshape(-1, 32*101)[0]
            test_data.append(temp)
            test_label.append(new_data_dic[i]['labels'][j])
print("Finished!")

process s06
process s18
process s27
process s07
process s15
process s10
process s08
process s02
process s31
process s16
process s28
process s14
process s32
process s23
process s09
process s22
process s29
process s24
process s13
process s04
process s19
process s01
process s12
process s20
process s11
process s26
process s03
process s21
process s25
process s30
process s17
process s05
Finished!


### Define decision tree model and train it

In [13]:
dt = tree.DecisionTreeClassifier(criterion='entropy')

# fit it to our data
dt.fit(train_data, train_label)

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [14]:
result = []
for i in range(len(test_data)):
    result.append(dt.predict(test_data[i].reshape(1,-1))[0])

count = 0
for index, i in enumerate(result):
    for j in range(len(i)):
        if(i[j]==test_label[index][j]):
            count += 1
print("The Accuracy is: {}%".format(count/(len(test_label)*4) * 100))

The Accuracy is: 49.375%


### Data Preprocessing
- Three classes

In [9]:
new_data_dic_3 = defaultdict(lambda:defaultdict(lambda:[]))
for key in data_dic.keys():
    par = int(key[1:])
    
    for i, vid in enumerate(data_dic[key]['data']):
        new_data_dic_3[key]['data'].append([])
        for chnn in vid[:-8]:
            l =preprocessing(np.array(chnn))
            for batch in range(0,10):
                l = l+preprocessing(np.array(chnn[batch*807:max((batch+1)*807, 8064)]))
            new_data_dic_3[key]['data'][i].append(np.array([par, i+1]+l))
    for i,  vid in enumerate(data_dic[key]['labels']):
        l2 = []
        new_data_dic_3[key]['labels'].append([])
        for value in vid:
            if value>=6:
                l2.append(2)
            if value<6 and value >= 4:
                l2.append(1)
            if value < 4:
                l2.append(0)
        new_data_dic_3[key]['labels'][i] += l2

In [20]:
train_data_3 = []
train_label_3 = []
test_data_3 = []
test_label_3 = []
for i in new_data_dic:
    print("process {}".format(i))
    if i != 's02':
        for j in range(len(new_data_dic_3[i]['labels'])):
            temp = np.array(new_data_dic_3[i]['data'][j]).reshape(-1, 32*101)[0]
            train_data_3.append(temp)
            train_label_3.append(new_data_dic_3[i]['labels'][j])
    else:
        for j in range(len(new_data_dic_3[i]['labels'])):
            temp = np.array(new_data_dic_3[i]['data'][j]).reshape(-1, 32*101)[0]
            test_data_3.append(temp)
            test_label_3.append(new_data_dic_3[i]['labels'][j])
print("Finished!")

process s06
process s18
process s27
process s07
process s15
process s10
process s08
process s02
process s31
process s16
process s28
process s14
process s32
process s23
process s09
process s22
process s29
process s24
process s13
process s04
process s19
process s01
process s12
process s20
process s11
process s26
process s03
process s21
process s25
process s30
process s17
process s05
Finished!


### Also, Train it

In [22]:
dt_3 = tree.DecisionTreeClassifier(criterion='entropy')
# fit it to our data
dt_3.fit(train_data_3, train_label_3)

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [24]:
result = []
for i in range(len(test_data_3)):
#     print(dt_3.predict(test_data_3[i].reshape(1,-1))[0])
    result.append(dt_3.predict(test_data_3[i].reshape(1,-1))[0])

count = 0
for index, i in enumerate(result):
    for j in range(len(i)):
        if(i[j]==test_label_3[index][j]):
            count += 1
print("The Accuracy is: {}%".format(count/(len(test_label_3)*4) * 100))

The Accuracy is: 30.625000000000004%


### Result
- Two classes classification Accuracy is : 49%~57%
- Three classes classification Accuracy is : 30%~42%